src/Simple.OData.Client.Core/ODataClient.Async.cs

```diff
@@ -1120,12 +1120,23 @@ internal async Task<T[]> ExecuteAsArrayAsync<T>(FluentCommand command, Cancellat
                 : result.Select(x => (T)x.ToObject(_session.TypeCache, typeof(T))).ToArray();
         }
 
-        internal async Task ExecuteBatchAsync(IList<Func<IODataClient, Task>> actions, CancellationToken cancellationToken)
+        internal async Task<ODataRequest> BuildRequestForBatch(IList<Func<IODataClient, Task>> actions, CancellationToken cancellationToken, IDictionary<string,string> headers = null)
         {
             await _session.ResolveAdapterAsync(cancellationToken).ConfigureAwait(false);
             if (cancellationToken.IsCancellationRequested) cancellationToken.ThrowIfCancellationRequested();
 
-            await ExecuteBatchActionsAsync(actions, cancellationToken).ConfigureAwait(false);
+            if (!actions.Any())
+                return null;
+
+            return await _lazyBatchWriter.Value.CreateBatchRequestAsync(this, actions, headers).ConfigureAwait(false);
+        }
+
+        internal async Task ExecuteBatchAsync(IList<Func<IODataClient, Task>> actions, CancellationToken cancellationToken, IDictionary<string, string> headers = null)
```

Comment:
Since this is internal, you should make `headers` non-optional and order it before the cancellation token.